### PR TITLE
[api] Support staking in Web3 

### LINKED ISSUE #2922: [api] Support other types of actions in Web3
**What would you like to be added**:

Support other types of action (like staking and rewarding) in web3
 - Transaction Object
 - Gas estimation
 - SendAction

### PR DESCRIPTION
**Pull request by Liuhaai**
fix #2922

### CODE REVIEW

File: action/rlp_tx_test.go [hunk 1]
```diff
@@ -90,185 +90,71 @@ func TestRlpDecodeVerify(t *testing.T) {
...
-			false,
-			92,
-			"9dcd5da20c43860f3a2ee099821949cf615942c1477b289dbef39c0d4087672f",
-			"04f79d369da4ae82d2e549f8ddc39cb38e1375434e595e5e5212d1db81ed7c7dbd319bb84f4654092de21f4f820e09467687f26a12420c1fca298f2a50f95455b5",
-			"47525b648d809d07c8878f9dd4b140724e3ce783",
+			"io1qqqqqqqqqqq8xarpdd5kue6rwfjkzar9k0wk6t",
+			356,
+			"773d70ddbe616cbe9343655baf06006213a2e31297cd11eccec4de3b7e35e81a",
+			"049c3a34c3f2053a769ed18f9266bc88e2404f3fc3c21fef75b561251d0af9898aa62e3fd0bfd5d4b0ac55cd912e200018fb8a396d29bf226962a082e8eec3cd20",
+			"64b690bd2db36de68c003692fe8c24b5dcf9e320",
```
Comment (dustinxie) [inline action/rlp_tx_test.go:157]: add test data for other staking actions

File: action/sealedenvelope.go [hunk 2]
```diff
@@ -130,52 +132,59 @@ func (sealed *SealedEnvelope) LoadProto(pbAct *iotextypes.Action) error {
...
+		ab = act.AbstractAction
+		data, err = act.EncodingABIBinary()
 	case *CandidateRegister:
-		tx, err = wrapStakingActionIntoExecution(act.AbstractAction, address.StakingRegisterCandAddrHash[:], act.Proto())
+		ab = act.AbstractAction
+		data, err = act.EncodingABIBinary()
 	case *CandidateUpdate:
-		tx, err = wrapStakingActionIntoExecution(act.AbstractAction, address.StakingUpdateCandAddrHash[:], act.Proto())
+		ab = act.AbstractAction
+		data, err = act.EncodingABIBinary()
```
Comment (dustinxie) [inline action/sealedenvelope.go:173]: can simplify to below?
```
case *Transfer, *Execution:
   return act, nil
case *CreateStake, *Unstake, *Restake ...:
    ab = act.AbstractAction
   data, err = act.EncodingABIBinary()
```

File: api/coreservice.go [hunk 3]
```diff
@@ -1490,3 +1487,34 @@ func (core *coreService) ReadContractStorage(ctx context.Context, addr address.A
 	}
 	return core.sf.ReadContractStorage(ctx, addr, key)
 }
+
+// TODO: use EstimateGasV2 in grpcserver.go
+// EstimateGasV2 estimates action gas except execution
+func (core *coreService) EstimateGasV2(actType action.Action, payloadSize uint64) (uint64, error) {
```
Comment (dustinxie) [inline api/coreservice.go:1493]: EstimateGasForNonExecution

File: api/web3server.go [hunk 4]
```diff
@@ -431,19 +431,48 @@ func (svr *Web3Server) estimateGas(in interface{}) (interface{}, error) {
 	}
 
 	var estimatedGas uint64
-	if isContract, _ := svr.isContractAddr(to); isContract {
+	actType, err := svr.actionTypeByAddr(to)
+	if err != nil {
+		return nil, err
+	}
```
Comment (dustinxie) [inline api/web3server.go:437]: var act action.Action

File: api/web3server.go [hunk 5]
```diff
@@ -431,19 +431,48 @@ func (svr *Web3Server) estimateGas(in interface{}) (interface{}, error) {
...
-		if err != nil {
-			return nil, err
-		}
-	} else {
-		// TODO: support staking actions
-		estimatedGas, err = svr.coreService.CalculateGasConsumption(action.TransferBaseIntrinsicGas, action.TransferPayloadGas, uint64(len(data)))
-		if err != nil {
-			return nil, err
+	case _transferAction:
+		estimatedGas, err = svr.coreService.EstimateGasV2(&action.Transfer{}, uint64(len(data)))
```
Comment (dustinxie) [inline api/web3server.go:443]: act := &action.Transfer{}

File: api/web3server.go [hunk 6]
```diff
@@ -431,19 +431,48 @@ func (svr *Web3Server) estimateGas(in interface{}) (interface{}, error) {
...
+	case _stakingAction:
+		if len(data) <= 4 {
+			return nil, errInvalidFormat
+		}
+		var methodID [4]byte
+		copy(methodID[:], data[:4])
+		switch methodID {
+		case action.CreateStakeMethodID:
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.CreateStake{}, uint64(len(data)))
```
Comment (dustinxie) [inline api/web3server.go:452]: actType := "stakeCreate"

File: api/web3server.go [hunk 7]
```diff
@@ -431,19 +431,48 @@ func (svr *Web3Server) estimateGas(in interface{}) (interface{}, error) {
...
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.TransferStake{}, uint64(len(data)))
+		case action.CandidateRegisterMethodID:
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.CandidateRegister{}, uint64(len(data)))
+		case action.CandidateUpdateMethodID:
+			estimatedGas, err = svr.coreService.EstimateGasV2(&action.CandidateUpdate{}, 0)
+		default:
+			return nil, errInvalidFormat
 		}
 	}
+	if err != nil {
```
Comment (dustinxie) [inline api/web3server.go:473]: estimatedGas, err = svr.coreService.EstimateGasForNonExecution(actType, len(data))

File: api/coreservice.go [hunk 8]
```diff
@@ -1490,3 +1487,34 @@ func (core *coreService) ReadContractStorage(ctx context.Context, addr address.A
...
+	var intrinsicGas, payloadGas uint64
+	switch act := actType.(type) {
+	case *action.Transfer:
```
Comment (dustinxie) [inline api/coreservice.go:1496]: case "transfer"

File: api/coreservice.go [hunk 9]
```diff
@@ -1490,3 +1487,34 @@ func (core *coreService) ReadContractStorage(ctx context.Context, addr address.A
...
+		intrinsicGas, payloadGas = action.TransferBaseIntrinsicGas, action.TransferPayloadGas
+	case *action.CreateStake:
```
Comment (dustinxie) [inline api/coreservice.go:1498]: case "stakeCreate"

File: action/sealedenvelope.go (hunk 2 above)
Comment (Liuhaai) [reply action/sealedenvelope.go]: No. There is a grammar error

File: action/envelope.go [hunk 10]
```diff
@@ -248,3 +250,51 @@ func (elp *envelope) ChainID() uint32 { return elp.chainID }
...
+// ToRLP converts native to RlpTransaction
+func (elp *envelope) ToRLP() (RlpTransaction, error) {
+	var (
+		err  error
+		data []byte
+		ab   AbstractAction
+	)
+	switch act := elp.payload.(type) {
+	case *Transfer:
+		return (*Transfer)(act), nil
```
Comment (dustinxie) [inline action/envelope.go:263]: return act, nil

File: action/envelope.go [hunk 11]
```diff
@@ -248,3 +250,51 @@ func (elp *envelope) ChainID() uint32 { return elp.chainID }
...
+	case *Execution:
+		return (*Execution)(act), nil
```
Comment (dustinxie) [inline action/envelope.go:265]: return act, nil

File: api/web3server.go (hunk 7 above)
Comment (dustinxie) [reply api/web3server.go:483]: yes, what i mean is `actType = "stakeCreate", "stakeUnstake"` for each different staking tx

File: api/web3server.go [hunk 12]
```diff
@@ -434,19 +434,55 @@ func (svr *Web3Server) estimateGas(in interface{}) (interface{}, error) {
...
-			return nil, err
+	case _transferAction:
+		estimatedGas, err = svr.coreService.EstimateGasForNonExecution(&action.Transfer{}, uint64(len(data)))
+	case _stakingAction:
+		if len(data) <= 4 {
+			return nil, errInvalidFormat
+		}
+		var err error
+		if act, err := action.NewCreateStakeFromABIBinary(data); err == nil {
+			estimatedGas, err = svr.coreService.EstimateGasForNonExecution(act, uint64(len(data)))
```
Comment (dustinxie) [inline api/web3server.go:453]:
```
if act, err := action.NewCreateStakeFromABIBinary(data); err == nil {
    actType = "stakeCreate"
    break
}
```

File: action/envelope.go [hunk 13]
```diff
@@ -248,3 +250,51 @@ func (elp *envelope) ChainID() uint32 { return elp.chainID }
...
+	case *CandidateUpdate:
+		ab = act.AbstractAction
+		data, err = act.EncodeABIBinary()
+	default:
+		return nil, errors.Errorf("invalid action type %T not supported", act)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return wrapStakingActionIntoExecution(ab, address.StakingCreateAddr, data)
```
Comment (dustinxie) [inline action/envelope.go:299]: we should file a new PR (in iotex-address repro) to define a new common address, and delete the current 9 special addresses

Comment (Liuhaai) [reply action/envelope.go:299]: oh, I will fix the special addr

Comment (Liuhaai) [reply action/envelope.go]: https://github.com/iotexproject/iotex-address/pull/19

File: api/web3server_utils.go [hunk 14]
```diff
@@ -306,19 +305,65 @@ func (svr *Web3Server) parseBlockRange(fromStr string, toStr string) (from uint6
...
 	accountMeta, _, err := svr.coreService.Account(ioAddr)
-	if err != nil {
-		return false, err
+	if err != nil && accountMeta.IsContract {
+		return _executionAction, nil
+	}
+	return _transferAction, nil
+}
+
+func loadStakingAction(data []byte, core *iotextypes.ActionCore) error {
```
Comment (dustinxie) [inline api/web3server_utils.go:326]: IMO action package is a better place for this code
web3 format transfer/execution are handled in there, web3 staking are also handled in the new `action.ToRLP()`, kinda weird to leave this part specifically in web3 package

File: action/rlp_tx.go [hunk 15]
```diff
@@ -115,3 +129,92 @@ func DecodeRawTx(rawData string, chainID uint32) (tx *types.Transaction, sig []b
 	pubkey, err = crypto.RecoverPubkey(rawHash[:], sig)
 	return
 }
+
+// EthTxExportToNativeProto converts eth tx to protobuf type ActionCore
+func EthTxExportToNativeProto(chainID uint32, actType int, tx *types.Transaction) (*iotextypes.ActionCore, error) {
```
Comment (dustinxie) [inline action/rlp_tx.go:134]: nit: EthTxToNativeProto()

File: api/web3server.go [hunk 16]
```diff
@@ -439,42 +439,42 @@ func (svr *Web3Server) estimateGas(in interface{}) (interface{}, error) {
...
+	case action.StakingActionType:
 		if len(data) <= 4 {
 			return nil, errInvalidFormat
 		}
 		var err error
 		if act, err := action.NewCreateStakeFromABIBinary(data); err == nil {
-			estimatedGas, err = svr.coreService.EstimateGasForNonExecution(act, uint64(len(data)))
+			estimatedGas, err = svr.coreService.EstimateGasForNonExecution(act, act.Payload())
 		}
 		if act, err := action.NewDepositToStakeFromABIBinary(data); err == nil {
```
Comment (dustinxie) [inline api/web3server.go:455]: any if is hit, no need to continue to run, so use
```
else if act, err := {}
else if act, err := {}
```

File: api/coreservice.go [hunk 17]
```diff
@@ -1531,8 +1531,9 @@ func (core *coreService) ReadContractStorage(ctx context.Context, addr address.A
 
 // TODO: use EstimateGasForNonExecution in grpcserver.go
 // EstimateGasForNonExecution estimates action gas except execution
-func (core *coreService) EstimateGasForNonExecution(actType action.Action, payloadSize uint64) (uint64, error) {
+func (core *coreService) EstimateGasForNonExecution(actType action.Action, payload []byte) (uint64, error) {
```
Comment (dustinxie) [inline api/coreservice.go:1534]: why change input type to `[]byte`? passing a `uint64` is more efficient than passing `[]byte`

**Review by dustinxie — APPROVED** [review]

**Comment by CoderZhi** [comment]
This PR is too big. Some of the logic in api package is hard to follow. Please split it into multiple PRs. You can start with staking action encoding/decoding part.

**Comment by Liuhaai** [comment]
> This PR is too big. Some of the logic in api package is hard to follow. Please split it into multiple PRs. You can start with staking action encoding/decoding part.

ok. Pls review this PR(https://github.com/iotexproject/iotex-core/pull/3092) first

**Comment by Liuhaai** [comment]
fixed by #3209  #3174  #3095
